### PR TITLE
Handle set-returning-function during pruning of computed columns [#113949429]

### PR DESCRIPTION
**Pull request by hsyuan**
@vraghavan78 @xinzweb @oarap 
Please take a look.

### CODE REVIEW

File: data/dxl/minidump/ArrayIN.mdp [hunk 1]
```diff
@@ -0,0 +1,876 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
```
Comment (vraghavan78) [inline data/dxl/minidump/ArrayIN.mdp]: why is this here? I think this got int accidentally by me. Please remove this.

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 2]
```diff
@@ -2131,8 +2131,21 @@ CTranslatorExprToDXL::PdxlnComputeScalar
 
 	// compute required columns
 	GPOS_ASSERT(NULL != pexprComputeScalar->Prpp());
+
 	CColRefSet *pcrsOutput = pexprComputeScalar->Prpp()->PcrsRequired();
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: can you please add a comment of what you are doing here?

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 3]
```diff
@@ -2131,8 +2131,21 @@ CTranslatorExprToDXL::PdxlnComputeScalar
...
 
+	ULONG ulPrLs = pexprProjList->UlArity();
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: const ulPrLs

File: libgpopt/src/translate/CTranslatorExprToDXL.cpp [hunk 4]
```diff
@@ -2131,8 +2131,21 @@ CTranslatorExprToDXL::PdxlnComputeScalar
...
+	for (ULONG ul = 0; ul < ulPrLs; ul++)
+	{
+		CExpression *pexprPrE = (*pexprProjList)[ul];
+		CDrvdPropScalar *pdpscalar = CDrvdPropScalar::Pdpscalar(pexprPrE->PdpDerive());
+		if (pdpscalar->FHasNonScalarFunction())
+		{
```
Comment (vraghavan78) [inline libgpopt/src/translate/CTranslatorExprToDXL.cpp]: why? Please add a comment

File: libgpopt/src/operators/CExpressionPreprocessor.cpp [hunk 5]
```diff
@@ -1666,10 +1666,14 @@ CExpressionPreprocessor::PexprPruneUnusedComputedCols
...
 	if (COperator::EopLogicalProject == pop->Eopid() || COperator::EopLogicalGbAgg == pop->Eopid())
 	{
-		pcrsReqdNew->Include(CLogical::PopConvert(pop)->PcrsLocalUsed());
 		CExpression *pexprProjList = (*pexpr)[1];
-		CColRefSet *pcrsUnusedLocal = GPOS_NEW(pmp) CColRefSet(pmp);
 		CColRefSet *pcrsDefined = CDrvdPropScalar::Pdpscalar(pexprProjList->PdpDerive())->PcrsDefined();
+		CColRefSet *pcrsSetReturningFunction = CDrvdPropScalar::Pdpscalar(pexprProjList->PdpDerive())->PcrsSetReturningFunction();
+
+		pcrsReqdNew->Include(CLogical::PopConvert(pop)->PcrsLocalUsed());
+		pcrsReqdNew->Union(pcrsSetReturningFunction);
```
Comment (vraghavan78) [inline libgpopt/src/operators/CExpressionPreprocessor.cpp]: please add a comment of why set running function need to union-ed

File: libgpopt/include/gpopt/base/CDrvdPropScalar.h [hunk 6]
```diff
@@ -112,6 +115,11 @@ namespace gpopt
 				return m_pcrsUsed;
 			}
 
+			CColRefSet *PcrsSetReturningFunction() const
```
Comment (vraghavan78) [inline libgpopt/include/gpopt/base/CDrvdPropScalar.h]: comment please for consistency

File: data/dxl/minidump/prune_set_returning_function.mdp [hunk 7]
```diff
@@ -0,0 +1,139 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
Comment (vraghavan78) [inline data/dxl/minidump/prune_set_returning_function.mdp]: Query?

File: data/dxl/minidump/prune_nested_subqueries.mdp [hunk 8]
```diff
@@ -0,0 +1,674 @@
+<?xml version="1.0" encoding="UTF-8"?><dxl:DXLMessage xmlns:dxl="http://greenplum.com/dxl/2010/12/">
+<dxl:Thread Id="0">
```
Comment (vraghavan78) [inline data/dxl/minidump/prune_nested_subqueries.mdp]: no ddl and query?

**Comment by vraghavan78** [comment]
+1

File: data/dxl/minidump/prune_nested_subqueries.mdp (hunk 8 above)
Comment (hsyuan) [inline data/dxl/minidump/prune_nested_subqueries.mdp]: yes, will remove it.

File: data/dxl/minidump/prune_set_returning_function.mdp (hunk 7 above)
Comment (hsyuan) [inline data/dxl/minidump/prune_set_returning_function.mdp]: will remove it.

File: data/dxl/minidump/ArrayIN.mdp (hunk 1 above)
Comment (hsyuan) [inline data/dxl/minidump/ArrayIN.mdp]: OK, will remove it.

File: libgpopt/src/operators/CExpressionPreprocessor.cpp (hunk 5 above)
Comment (hsyuan) [inline libgpopt/src/operators/CExpressionPreprocessor.cpp]: I didn't see any difference between union and include here. @xinzweb

**Comment by hsyuan** [comment]
updated according to the feedback. please take a look again.

File: libgpopt/src/base/CDrvdPropScalar.cpp [hunk 9]
```diff
@@ -196,6 +211,7 @@ CDrvdPropScalar::OsPrint
 {
 		os	<<	"Defined Columns: [" << *m_pcrsDefined << "], "
 			<<	"Used Columns: [" << *m_pcrsUsed << "], "
+			<<  "Set Returning Function Columns: [" << *m_pcrsSetReturningFunction << "], "
```
Comment (xinzweb) [inline libgpopt/src/base/CDrvdPropScalar.cpp]: space vs. tab

File: server/src/unittest/gpopt/minidump/CICGTest.cpp [hunk 10]
```diff
@@ -49,7 +49,6 @@ const CHAR *rgszFileNames[] =
 		"../data/dxl/minidump/OrderByNullsFirst.mdp",
 		// disabling test because it is non-deterministic.
```
Comment (xinzweb) [inline server/src/unittest/gpopt/minidump/CICGTest.cpp]: Can we just remove the deadcode?

File: server/src/unittest/gpopt/minidump/CPruneColumnsTest.cpp [hunk 11]
```diff
@@ -0,0 +1,106 @@
...
+//
+//	@owner:
+//
+//
+//	@test:
+//
+//
+//---------------------------------------------------------------------------
+
+#include "unittest/gpopt/minidump/CPruneColumnsTest.h"
```
Comment (xinzweb) [inline server/src/unittest/gpopt/minidump/CPruneColumnsTest.cpp]: sort the include.

**Comment by xinzweb** [comment]
+1 besides some minor formatting
